spring-cloud-gcp-logging/src/main/java/org/springframework/cloud/gcp/logging/CompositeTraceIdExtractor.java

```diff
@@ -0,0 +1,57 @@
+/*
+ *  Copyright 2018 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.logging;
+
+import java.util.List;
+
+import javax.servlet.http.HttpServletRequest;
+
+import com.google.common.collect.ImmutableList;
+
+import org.springframework.util.Assert;
+
+/**
+ * Checks HTTP request for multiple headers that might contain a trace ID, and provides
+ * the first one.
+ *
+ * @author Chengyuan Zhao
+ */
+public class CompositeTraceIdExtractor implements TraceIdExtractor {
```

Comment:
It seems to me like this whole trace ID extraction mechanism can be resumed to one or two private methods.
When will we need to add more headers that justify this extensibility overhead?
Even if we do, it should be as simple as adding a new string to a Set constant or so...

Reply:
It was very simple to start, and in fact the existing code is like that, but we had alot of offline discussion and we settled on this approach.